### PR TITLE
Fix pickle support for DataFrame and Series

### LINKED ISSUE #672: Modin objects are not able to be pickled
```
File ".../site-packages/cache_memoize/__init__.py", line 116, in inner
    result = cache.get(cache_key, MARKER)
  File ".../site-packages/django_redis/cache.py", line 32, in _decorator
    return method(self, *args, **kwargs)
  File ".../site-packages/django_redis/cache.py", line 81, in get
    client=client)
  File ".../site-packages/django_redis/client/default.py", line 210, in get
    return self.decode(value)
  File ".../site-packages/django_redis/client/default.py", line 318, in decode
    value = self._serializer.loads(value)
  File ".../site-packages/django_redis/serializers/pickle.py", line 35, in loads
    return pickle.loads(value)
  File ".../site-packages/modin/pandas/base.py", line 3337, in __getattribute__
    if self.empty and is_callable:
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)
  File ".../site-packages/modin/pandas/dataframe.py", line 211, in empty
    return len(self.columns) == 0 or len(self.index) == 0

  File ".../site-packages/modin/pandas/dataframe.py", line 131, in _get_columns
    return self._query_compiler.columns
  File ".../site-packages/modin/pandas/dataframe.py", line 1898, in __getattr__
    if key in self.columns:
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)

#  ... So many times!

  File ".../site-packages/modin/pandas/dataframe.py", line 131, in _get_columns
    return self._query_compiler.columns
  File ".../site-packages/modin/pandas/base.py", line 3343, in __getattribute__
    return object.__getattribute__(self, item)
RecursionError: maximum recursion depth exceeded while calling a Python object
```
 I use django-cache to cache the result of modin df, When django want to fetch the cache,It hit this issue.

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
Implement fine-grained control over how `DataFrame` and `Series` objects are pickled and restored.
Main idea is to really pickle Pandas counterparts and restore Modin objects from them upon unpickling.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #672
- [x] tests added and passing

### CODE REVIEW

**Comment by vnlitvinov** [comment]
I've converted to draft because I think I might be pickling entire dataframe (i.e. too much data) instead of single column.

**Comment by anmyachev** [comment]
@vnlitvinov why you removed `__getstate__` and `__setstate__`?  Also please add docstring for `__inflate__` and `__reduce__` methods.

File: modin/pandas/dataframe.py [hunk 1]
```diff
@@ -3121,8 +3121,12 @@ def __contains__(self, key):
     def __round__(self, decimals=0):
         return self._default_to_pandas(pandas.DataFrame.__round__, decimals=decimals)
 
-    def __setstate__(self, state):
-        return self._default_to_pandas(pandas.DataFrame.__setstate__, state)
+    @classmethod
+    def __inflate__(self, pandas_df):
+        return from_pandas(pandas_df)
```
Comment (devin-petersohn) [inline modin/pandas/dataframe.py:3126]: This seems like it will have serious performance issues, is this the only way?

**Review by devin-petersohn — COMMENTED** [review]
For my understanding I have a few questions:

1. Is this intended to support pickling between Ray drivers? Sending data between Ray clusters?
2. Ray `ObjectRef`s are picklable, so we should just be able to pickle the `query_compiler` object and re-wrap that in `modin.pandas.DataFrame` on the receiving end
3. It might be useful to have two separate ways of pickling

**Comment by vnlitvinov** [comment]
@anmyachev 

> why you removed `__getstate__` and `__setstate__`?

They're providing more high-level pickling support... I'm implementing the lowest portion of the pickle protocol, as Modin objects play too much with `__getattribute__()`, so it's pretty hard to initialize them using `__setstate__` - it requires `__getattribute__` to work before it could be properly used, and we cannot make it work without initializing the object (kind of closed loop).

@devin-petersohn 

> 1. Is this intended to support pickling between Ray drivers? Sending data between Ray clusters?

Main reason _for me_ is to enable remote context, where in certain functions we pass objects "by value" which means we pickle them locally and unpickle remotely or vice versa. Hence to pass Dataframes or Series we need them to be picklable :) 


> 3. It might be useful to have two separate ways of pickling

We might even want to have each IO class define its own pickling ways in the end. This is (probably) just the first step, probably inefficient but still better than crashing with "Recursion error".

File: modin/pandas/dataframe.py (hunk 1 above)
Comment (vnlitvinov) [reply modin/pandas/dataframe.py:3126]: I don't think it's the only way, my original implementation just gave away `self._query_compiler` as a thing to pickle and then reconstructed `DataFrame` from it. The reason I chose this particular one is I wanted to make sure the ending pickle is persistent enough to be storable on disk and readable in a separate process run, for which I'm not sure storing Ray objects or whatever is what we need.

I'm also keeping in mind my goal of enabling local-remote context talk, and I don't think I can pickle local Ray object and unpickle it remotely.

Maybe we need a switch of "slow but storable" and "fast but works in one session only"?

**Comment by vnlitvinov** [comment]
Superseded by #2835

**Pull request closed without merging**